frontend/public/components/utils/name-value-editor.jsx

```diff
@@ -113,3 +100,129 @@ NameValueEditor.defaultProps = {
+    // When dragging downwards, only move when the cursor is below 50%
+    // When dragging upwards, only move when the cursor is above 50%
+
+    // Dragging downwards
+    if (dragIndex < hoverIndex && hoverClientY < hoverMiddleY) {
+      return;
+    }
+
+    // Dragging upwards
+    if (dragIndex > hoverIndex && hoverClientY > hoverMiddleY) {
+      return;
+    }
+
+    // Time to actually perform the action
+    props.onMove(dragIndex, hoverIndex);
+
+    // Note: we're mutating the monitor item here!
+    // Generally it's better to avoid mutations,
+    // but it's good here for the sake of performance
+    // to avoid expensive index searches.
+    monitor.getItem().index = hoverIndex;
+  }
+};
+
+const collectSourcePair = (connect, monitor) => {
+  return {
+    connectDragSource: connect.dragSource(),
+    connectDragPreview: connect.dragPreview(),
+    isDragging: monitor.isDragging()
+  };
+};
+
+const collectTargetPair = (connect) => {
+  return {
+    connectDropTarget: connect.dropTarget(),
+  };
+};
+
+const PairElement = DragSource(DRAGGABLE_TYPE.ROW, pairSource, collectSourcePair)(DropTarget(DRAGGABLE_TYPE.ROW, itemTarget, collectTargetPair)(class PairElement extends React.Component {
+  constructor(props) {
+    super(props);
+  }
+
+  render() {
+    const {isDragging, connectDragSource, connectDragPreview, connectDropTarget, nameString, valueString, allowSorting, readOnly, index, onChange, onRemove, pair} = this.props;
+    const deleteButton = <span><i className="fa fa-minus-circle pairs-list__btn pairs-list__delete-icon" aria-hidden={true} onClick={() => onRemove(index)}></i><span className="sr-only">Delete</span></span>;
+
+    return connectDropTarget(
+      connectDragPreview(
+        <div className={classNames('row', isDragging ? 'pairs-list__row-dragging' : 'pairs-list__row')} ref={node => this.node = node}>
+          <div className="col-xs-5 pairs-list__field">
+            <input type="text" className="form-control" placeholder={nameString.toLowerCase()} value={pair[NameValueEditorPair.Name]} onChange={e => onChange(e, index, NameValueEditorPair.Name)} disabled={readOnly}/>
+          </div>
+          <div className="col-xs-6 pairs-list__field">
+            {
+              _.isPlainObject(pair[NameValueEditorPair.Value]) ?
+                <ValueFromPair pair={pair[NameValueEditorPair.Value]}/> :
+                <input type="text" className="form-control" placeholder={valueString.toLowerCase()} value={pair[NameValueEditorPair.Value] || ''} onChange={e => onChange(e, index, NameValueEditorPair.Value)} disabled={readOnly}/>
+            }
+          </div>
+          {readOnly ? null :
+            <div className="col-xs-1">
+              <span className={classNames(allowSorting ? 'pairs-list__span-btns' : null)}>
+                {allowSorting ? <span>{connectDragSource(<i className="fa fa-bars pairs-list__reorder-icon"/>)}{deleteButton}</span>:
+                  <span>{deleteButton}</span>
```

Comment:
Prefer `<React.Fragment>` to spans with no attributes

Do you mind cleaning up the formatting some here? It would be easier to follow the logic if split across lines.

Reply:
No problem. made it easier on the 👁 's. Used <React.Fragment> as well.